# Gen-ie

Gen-ie makes early diagnosis magic for detecting rare diseases to prevent the patient from going on a long journey to find the diagnosis. Gen-ie can significantly decrease the journey of diagnosing the disease via genomic sequencing. Our sequencing engine can detect and sequence 500 rare disease genes. Hence, it pinpoints how the specific mutated nitrogenous base found in DNA leads to disease development and processes the whole human genome, which has 3 billion genes. A national health system used whole-genome sequencing (WGS) to find new genetic causes and enhance uncommon disease diagnosis. Out of 13,037 participants, 1,138 out of 7,065 extensively phenotyped people received a genetic diagnosis, including 9,802 with uncommon disorders. Of the 95 Mendelian gene-disease connections found, 11 were found after 2015, and 79 were shown to be causal.
Additionally, WGS identified four new non-coding variations contributing to disease by interfering with ARPC1B, GATA1, LRBA, and MPL. Data analysis from the UK Biobank also revealed that uncommon alleles influence extreme red blood cell characteristics. This study advances the study of uncommon diseases in clinical settings by highlighting the effectiveness of WGS in diagnostic and etiological discovery. (Turro et al., 2020).
For the data analysis, we have used Clingen API using individualized research towards the protein and mutated gene where the used parameters like Gene region of causation of the mutation specifically chromosome, Haploinsufficiency, Triplosensitivity, and hereditary types of disease (e.g. X-linked, Autosomal Dominant). The created framework provides a scoring matrix (1-10) and color gradient. The matrix was created from data analyzed via Clingen API. Unique values were assigned to the categories, such as the relevance of the research on the disease using a metric system of standard operating procedures (SOPs). A higher SOP number means higher relevance. Also, for the relevance calculation, we used classification, which had values like definitive (1.0), strong (0.9), medium (0.5), weak (0.3), and no known disease (0.05), disputed (0.05), and refuted (0.1) Also, there is sorting the gene mutation to their mechanism of the mutations like function Lost Of Function (LOF), Gain Of Function (GOF), and Dominant Negative (DN) so they represented as columns of the matrix. So during analysis using the UniProt, we found that positive Haploinsufficiency and some hereditary inheritance like the X-linked recessive and autosomal recessive correlates towards LOF; for the GOF, we had a correlation with Autosomal dominants, positive Triplosensitivity, and chromosome region for the mutation was chr 2,3,6,12,13,14,17,19. However, during the dataset analysis, it had missing lines of information, or the disease was not clearly researched by clinical trials; therefore, with method elimination, the rest of the gene mutations were DN. 
Polyphen-2 was also included and used to predict the potential impact of an amino acid substitution on the structure and function of human proteins using simple physical and comparative considerations. We used Excel for gene mutation categorization of the data, PubMed for the research of data, UniProd individualized analysis of proteins and genes, Google Scholar for the academic paper search, and NIH cross-link protein function and location overview of proteins and genes. The estimated 70% efficiency of the sorting of the gene data function was estimated, and the rest was insufficient data to increase. 

![image](https://github.com/user-attachments/assets/fce4ef63-0f96-4518-b28f-94c4ff828ee2)
